FIPS/fip-draft_foundational_fvm.md

```diff
@@ -0,0 +1,1010 @@
+  emitted by the actor or the runtime, and they're recorded on-chain inside the
+  receipt.
+- **Syscall error numbers:** signal the outcome of a single syscall. They are
+  ephemeral, and have no meaning beyond the FVM syscall boundary.
+
+In both cases, a `0` value implies success. For more information, refer to the
+[FVM Error Conditions] specification (commit `724650` at the time of writing
+this FIP).
+
+### Memory policy
+
+The Invocation Container makes available a maximum of 1024 pages to the actor.
+Given that WASM memory pages are a fixed 64KiB size, this results in every actor
+call being limited to 64MiB of memory.
+
+Depending on the WASM engine, it may be possible to size memory dynamically. If
+that's the case, a policy that grows memory dynamically from a low value (even
+zero) should be preferred.
+
+### Module caching
+
+WASM execution performance is a determining factor in the ability to maintain
+blockchain epoch time (30 seconds).
+
+We recommend implementations to compile WASM bytecode into machine code for
+faster execution, instead of relying on interpretation. This incurs in an
+upfront CPU-bound compilation cost, in exchange for ongoing operating costs.
+
+Actively or passively-managed module caches could be explored, but such
+approaches may introduce vulnerabilities. Instead, we may factor in the
+compilation compute cost and the perpetual storage cost, as deployment-time gas.
+
+Any approach will need to account for eventual recompilation costs on events
+like WASM Engine or compiler upgrades.
+
+As far as the network upgrade being proposed in this FIP is concerned, we
+recommend implementations to warm up the module cache for all built-in actors
+during node initialization.
+
+### Gas charging
+
+The current gas model relies on actor code being static and known ahead of time.
+Gas is currently charged at the syscall level and not at the instruction level.
+
+With the ability to run arbitrary code on the network, the gas model must
+account for actual execution costs at the instruction level, in addition to
+syscall-level charges.
+
+A gas remodelling proposal will follow this FIP.
+
+### Specific actor coupling
+
+Due to how certain parts of the Filecoin protocol work, implicit coupling that
+exists between the VM and specific actor types will be carried over. Concretely:
+
+- Account actor: the FVM must create new account actors when processing sends to
+  pubkey addresses.
+- Init actor: the FVM must load and query the init actor state when resolving
+  addresses during message processing, and when the `actor::resolve_address` syscall is
+  called. Additionally, a new init actor entry must be added when implicitly
+  creating an account, as per above.
+- Power actor and Storage Market actor: state is queried when handling the
+  `network::total_fil_circ_supply` syscall.
+- Burnt funds account and Reserve account: balances are queried when handling
```

Comment:
Please lmk if there are any change we could make here that would make things significantly easier for you.